### DISCUSSION TITLE
A Quantitative Comparison of Serving Platforms for Neural Networks

### DISCUSSION THREAD

**Discussion by illagrenan** [Biano-AI/biano-ai.github.io#3] [General]
Comments to [the article](https://biano-ai.github.io/research/2021/08/16/quantitative-comparison-of-serving-platforms-for-neural-networks.html).

**Comment by illagrenan** [comment]
A link to the article was uploaded to Reddit: https://www.reddit.com/r/mlops/comments/p7ehb7/a_quantitative_comparison_of_serving_platforms/

**Comment by Bunoviske** [comment]
Hello,

great blog article! My question is: how can I authorize users before they access the serving engine endpoint? What are the common practices on that?

Thank you!

**Reply by illagrenan** [reply]
Our testing middleware is written in FastAPI, which supports user authentication: https://fastapi.tiangolo.com/tutorial/security/.

Another option would be to put a reverse proxy in front of the serving, which would be used for authentication. Some examples of such proxies:

* https://www.nginx.com/, https://docs.nginx.com/nginx/admin-guide/security-controls/configuring-http-basic-authentication/
* For Kubernetes: https://kubernetes.github.io/ingress-nginx/examples/auth/basic/
* ...

**Reply by Bunoviske** [reply]
Thank you! So how do you put FastAPI on top of the serving engine? Do you use microservices? Then your serving engine is only accessible by the FastAPI which has authentication?

**Reply by illagrenan** [reply]
The FastAPI code is here: https://github.com/Biano-AI/serving-compare-middleware And yes, servings are not accessible from the outside - only via FastAPI.

The following is a diagram of an architecture that we have been using in production for several years and that has worked well for us:
![Servings--Architecture](https://user-images.githubusercontent.com/320373/178259956-5a7ff8b3-baf2-43af-a43f-0d5963ecd87b.png)

**Reply by Bunoviske** [reply]
That is really helpful and clarifying, thank you. I was always trying to avoid one extra API call during inference. In other words, I though that the client acessing directly the serving endpoint would be more performant. 

One last question: you did not put in your comparison inference with ONNX, without using any engines. Why would I prefer to setup these serving engines instead of just using ONNX directly or even Torchscript? The serving engines are much more performant?

**Reply by illagrenan** [reply]
Yes, you're right. Once you put another component between the client and the model, it has an impact on performance. In our case, it's tens of milliseconds at most, which we're OK with.

I don't know about the ONNX stuff, let's ask my colleague @toretak who is an expert in this area.

**Reply by toretak** [reply]
@Bunoviske Hello, thanks for your comment!

To your question: ONNX is standard for model representation. ONNX runtime is more model optimization.
We just want to test the official servings from the most famous frameworks (or better to say, frameworks we use, because there are many more) against Triton.
Triton and Torchserve can both run ONNX models. It would be interesting to test them against ONNX runtime.  👍

**Comment by SimZhou** [comment]
There must be something wrong.
It is impossible that TF-serving is 40x slower than Triton.
In our practical experiences, tf-serving is pretty reliable and well-performed.

**Reply by toretak** [reply]
Hi,
sure, it is possible. You can clone repo, replicate the procedure, and find what's wrong. We will appreciate it.

In fact, from our experiences, it's true because we had quite a lot of performance issues while we used TFserving. Since we use Torchserve and Triton we are able to process much more data with less stress...

Thanks for your opinion.

**Comment by pjh4993** [comment]
I have a question about results reported on this article. 

For torchserve, Did you checked job queue size of server? In here, job queue refers to scheduling queue for request handler. Default max job queue size is set to 100, which can incur large fail rate when concurrency increases. TRIS(triton inference server), on the other hand doesn't restrict maximum job queue size.
